src/base/rss/rss_autodownloader.cpp

```diff
@@ -0,0 +1,374 @@
+#include "rss_session.h"
+
+struct ProcessingJob
+{
+    QString feedURL;
+    QString articleGUID;
+    QString articleTitle;
+    QString torrentURL;
+};
+
+const QString DataFolder(QStringLiteral("rss/autodownloader"));
+const QString RulesFile(QStringLiteral("rules.json"));
+
+const QString SettingsKey_ProcessingEnabled(QStringLiteral("RSS/AutoDownloader/EnableProcessing"));
+
+using namespace RSS;
+
+QPointer<AutoDownloader> AutoDownloader::m_instance = nullptr;
+
+AutoDownloader::AutoDownloader()
+    : m_processingEnabled(SettingsStorage::instance()->loadValue(SettingsKey_ProcessingEnabled, false).toBool())
+    , m_processingTimer(new QTimer(this))
+    , m_ioThread(new QThread(this))
+{
+    Q_ASSERT(!m_instance); // only one instance is allowed
+    m_instance = this;
+
+    m_fileStorage = AsyncFileStorage::create(
+                Utils::Fs::expandPathAbs(specialFolderLocation(SpecialFolder::Data) + DataFolder));
+    if (!m_fileStorage)
+        throw std::runtime_error("Directory for RSS AutoDownloader data is unavailable.");
+
+    m_fileStorage->moveToThread(m_ioThread);
+    connect(m_ioThread, &QThread::finished, m_fileStorage, &AsyncFileStorage::deleteLater);
+    connect(m_fileStorage, &AsyncFileStorage::failed, [](const QString &fileName, const QString &errorString)
+    {
+        Logger::instance()->addMessage(QString("Couldn't save RSS AutoDownloader data in %1. Error: %2")
+                                       .arg(fileName).arg(errorString), Log::WARNING);
+    });
+
+    m_ioThread->start();
+
+    connect(BitTorrent::Session::instance(), &BitTorrent::Session::downloadFromUrlFinished
+            , this, &AutoDownloader::handleTorrentDownloadFinished);
+    connect(BitTorrent::Session::instance(), &BitTorrent::Session::downloadFromUrlFailed
+            , this, &AutoDownloader::handleTorrentDownloadFailed);
+
+    load();
+
+    m_processingTimer->setSingleShot(true);
+    connect(m_processingTimer, &QTimer::timeout, this, &AutoDownloader::process);
+
+    if (m_processingEnabled) {
+        foreach (Article *article, Session::instance()->rootFolder()->articles()) {
+            if (!article->isRead() && !article->torrentUrl().isEmpty())
+                addJobForArticle(article);
+        }
+        connect(Session::instance()->rootFolder(), &Folder::newArticle, this, &AutoDownloader::handleNewArticle);
+    }
+}
+
+AutoDownloader::~AutoDownloader()
+{
+    store();
+
+    m_ioThread->quit();
+    m_ioThread->wait();
+}
+
+AutoDownloader *AutoDownloader::instance()
+{
+    return m_instance;
+}
+
+bool AutoDownloader::hasRule(const QString &ruleName) const
+{
+    return m_rules.contains(ruleName);
+}
+
+AutoDownloadRule AutoDownloader::ruleByName(const QString &ruleName) const
+{
+    return m_rules.value(ruleName, AutoDownloadRule("Unknown Rule"));
+}
+
+QList<AutoDownloadRule> AutoDownloader::rules() const
+{
+    return m_rules.values();
+}
+
+void AutoDownloader::setRule(const AutoDownloadRule &rule)
```

Comment:
When it comes to collections, I agree `insert` is correct name. But AutoDownloader is not a collection of rules, it just contains a collection of rules, so `insert` here is incorrect to. But since I can't find a more appropriate name, I change it on `insertRule`, since it is more like you. Although, maybe `useRule` will be better?